basic_pipeline/6_OrderingBuffer.md

```diff
@@ -0,0 +1,171 @@
+```
+
+We simply send the `:demand` on the `:input` pad once we receive a demand on the `:output` pad. Packets are not aggregated in any way so for each 1 unit of demand we send 1 unit on demand to the `:input` pad.
+
+Now we can go to the main part of the Ordering Buffer implementation - the `handle_process/4` callback:
+```Elixir
+# FILE: lib/elements/OrderingBuffer.ex
+
+defmodule Basic.Elements.OrderingBuffer do
+  ...
+  @impl true
+  def handle_process(:input, buffer, _context, state) do
+    packet = unzip_packet(buffer.payload)
+    ordered_packets = [packet | state.ordered_packets] |> Enum.sort()
+    state = Map.put(state, :ordered_packets, ordered_packets)
+    {last_seq_id, _} = Enum.at(ordered_packets, 0)
+    ...
+  end
+
+  defp unzip_packet(packet) do
+    regex = ~r/^\[seq\:(?<seq_id>\d+)\](?<data>.*)$/
+    %{"data" => data, "seq_id" => seq_id} = Regex.named_captures(regex, packet)
+    {String.to_integer(seq_id), data}
+  end
+  ...
+end
+```
+
+First, we are taking advantage of the [Regex module](https://hexdocs.pm/elixir/1.13/Regex.html) available in the Elixir.
+With the `Regex.named_captures` we can access the values of the fields defined within the regex description.
+Do you remember what our packet looks like? 
+```
+[seq:7][frameid:2][timestamp:3]data
+```
+Above you can see an exemplary packet. We need to fetch the value of the sequence id (in our case it is equal to 7) and get the rest of the packet.
+Therefore we have defined the regex description as:
+```Elixir
+~r/^\[seq\:(?<seq_id>\d+)\](?<data>.*)$/
+```
++ `~r/.../` stands for the `sigil_r/1` [sigil](https://elixir-lang.org/getting-started/sigils.html)
++ `^` describes the beginning of the input
++ `\[` stands for the opening square bracket ('[') at the beginning is required to escape the char since the plain '[' has a special meaning in the regex syntax
++ `seq` is a sequence of 's', 'e', 'q' characters (we need to adjust our regex description to match the header of the packet)
++ `\:` stands for the ':' character (we also need to escape that character since it is meaningful in the regex's syntax)
++ `(?<seq_id>\d+)` allows us to define a named capture - later one, once we use the `Regex.named_captures/2`, we will retrieve the map with 'seq_id' key and the corresponding value equal to the string described by the `\d+` 'partial' regex (which means - one or more occurrences of a decimal). Generally speaking, a named capture can be specified with the following structure: `(?<key>regex)` where `regex` is a regex description.
++ `\]` is the escaped closing square bracket character
++ `(?<data>.*)` is a named capture description that allows us to get the value of a `.*` regex (any character no or any number of times) under a `data` key.
++ `$` stands for the end of the input
+
+The result of `Regex.named_captures/2` applied to that regex description and the exemplary packet should be following:
+```Elixir
+{"seq_id"=>7, "data"=>"[frameid:2][timestamp:3]data"}
+```
```

Comment:
Hmm, I am just describing our particular regex, not regexes generally. I think it might be worth leaving it that way as the pure regex looks somehow complicated.